### ISSUE TITLE
KeepassXC and Firefox KPXC addon, flaky initial connecting operation on a test system (windows 10, 64 bit v1909)

### ISSUE THREAD

**Issue by succeedmr123** [keepassxreboot/keepassxc#4666]
## Overview
[TIP]:  # ( DO NOT include screenshots of your actual database! )
[NOTE]: # ( Give a BRIEF summary about your problem )

KeepassXC 2.5.4+Firefox KeepassXC Addon 1.6.2 (26-03-2020) = Flaky operation under w10 1909 x64

a.) main bug manifestation: Firefox KPXC addon cannot connect to external KeepassXC app without resorting to a workaround to start working (in Firefox - KPXC addon off-on under Addons menu)
 
b.) Scenario 2: it fails more prominently under stress test of Firefox browser itself (300 opened tabs and multiple windows in a research environment)
                                                                                                 |
                                                                                                \/
## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )

0. Preinstall: Install KeepassXC for windows 10 v1909 64-bit
Install Firefox Addon KeepassXC 1.6.2 (26-03-2020) 

1.  try to open keepassXC via Firefox KeepassXC Addon (it stopped working on my system a few months ago)
2.  You might experience that KeepassXC can not be opened, and you will have to open it manually - unlike beforehand some versions ago.
3.  You will witness that the addon will absolutely refuse to work properly, unless applying a trivial, but flow-breaking, uncomfortable workaround of disabling and re-enabling it (the necesssity to interrupt your productivity flow, and deal with addon disable/enable in firefox "addons" upon starting the KeepassXC and loading the database)

## Expected Behavior
[NOTE]: # ( Tell us what you expected to happen )

Fast,lean, reliable operation, as envisioned

## Actual Behavior
[NOTE]: # ( Tell us what actually happens )

It refuses to work properly from the get-go, and second bug exhibits under firefox stress test ("check status" goes to neverending waiting). It was, and still obviously, acts very flaky.

(That was the reason why I personally abandoned original Keepass plugin, and an  app as a daily driver, (besides the necessity to make Keepass work in Linux via Mono additional setup and tweaking)

After the extremely long time dealing with Firefox, and all the experiences with addons+FF behaviours, got my own theory why it happens with password filler addons, but for the sake of clarity I leave it out of this report.

## Context
[NOTE]: # ( Give us any additional information you may have. )

The problem is additionally pronounced under heavy amount of windows and multiple tabs in each Firefox window.
In that scenario,  even if you do a quick disable-re-enable keepassXC addon in firefox under Addons, the addon will refuse to work properly, but this time waiting infinitely with "checking status..."

[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - Version 2.5.4
Revision: c023bf9

Qt 5.14.2
Debugging mode is disabled.

Operating system: Windows 10 Version 1909
CPU architecture: x86_64
Kernel: winnt 10.0.18363

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey

Cryptographic libraries:
 libgcrypt 1.8.5

[NOTE]: # ( Pick choices based on your environment )
Operating System: Windows 10 64 bit
Desktop Env: Win explorer.exe DE
Windowing System: Windows 10 native

**Comment by droidmonkey** [comment]
I cannot follow what you are trying to say. Does this behavior only happen when you have 300 tabs open in Firefox?

**Comment by varjolintu** [comment]
Does it change anything if you disable the _Use dynamic input field detection_  from extension's settings?

**Comment by succeedmr123** [comment]
@varjolintu - thanks, will test and report back

@droidmonkey 
Post-mortem analysis;

Shortest posssible description; regardless of one or literally 366 Firefox tabs at this system -  KeepassXC addon under Firefox is, unfortunately, acting unreliably for now. Current issue is noticed for over few months period.

Will try it out with setting suggested by @varjolintu now

**Comment by varjolintu** [comment]
I'm also interested what `about:performance` shows.

**Comment by succeedmr123** [comment]
Progress report (from an average user perspective);

In KPXC FF under addon Options/General;

-as per @varjolintu  suggestion unticked  "Use dynamic input field detection" 

and , additionally,
-unticked experimental feature "Automatically reconnect to KeePassXC"

Under these conditions and 6 tabs opened, few scenarios were probed and results are as follows;

1. KeepassXC native app started and db login made before Firefox -> KPXC Addon connects successfully

2. Firefox started without the KPXC native app -> the KPXC Addon fails to connect (predictably), then suggests to  check is KPXC main app running

3. FF started without the KPXC , then KPXC native app started and db login -> Addon displays brief connected icon, then swiftly chang it for not connected icon, when clicked with mouse it shows a KPXC addon Error: "cannot decrypt message". Upon subsequent mouse click, error change to "Key exchange not successful"

When I will have more room for testing, will report back here.

Best,

**Comment by succeedmr123** [comment]
2020-05-02_22h Further Test:

Same setup;
FF Multi window
FF Multi tabs
same addons as the test up:
KPXC native is not logged into db but started hours ago

The problem manifestation  + how to repeat;

Manifest: Same - it refuses connection even when you login into native KPXC app db again.

That is under sceanario 2. from the test reported beforehand, and as the changes were made in the report beforehand 
( In short;
-FF is active
-you additionally after opening tabs etc within FF - start KeepassXC native App
-then, trying to establish connection within FF by pressing the KeepassXC Addon icon)

After that,

temporary workaround: Go to Firefox / Addons / KeepassXC Addon and do an Off - On toggle, the KPXC addon only then starts checking, and establish connection in Firefox successfully.

My thoughts:
link and signaling between KPXC and KPXC FF Addon is inconsistent. FF Addon does not get the signal when you login to DB, and pressing the icon fails to connect KPXC and Addon.

Which factors come to play, I could only speculate but I refrain from doing that.

Best,

**Comment by varjolintu** [comment]
- What about the `about:performance`?
- Have you tried to create a new temporary profile to Firefox and see if the same happens there?

The addon gets a signal when DB is locked only if there's a connection between them, but as you said, you have problems starting it. Just pressing the icon does not reconnect the extension. You need to use the "Reload" button if Auto-Reconnect is disabled.

**Comment by succeedmr123** [comment]
@varjolintu 

sure man, let's do that;

but , first to refresh on what changed in test env today:

-FF version upped to 76.0
-Under **Scenario 3** (defined few posts ago as the test environment cases) -> Still the error "checking status" -> "cannot decrypt message" ["Reload" pressed] -> Key exchange was not successful 

Now switching to a clean test profile and report back here in a separate post.

Best,




 can not

**Comment by succeedmr123** [comment]
@varjolintu 

Tested with "empty" test profile.

Result:

it works under these conditions flawlessly.

that leads me to believe that digging deeper in production profile is the answer - cause disabling key addons is not preferable at all;

e.g. in production profile;
-Containers are enabled as functionality,
-the meta addons that rely on container functionality such as 
-auto open in container - main tracking domains/sites such as google..facebook..twitter..linkedin..amazon
-then we've got umatrix as tracker manager, etc

To be clear, I am stating this as an use case and decisions for my production setup, am willing to dig into debugging, and am not willing to lose functionality of enhanced privacy and anti-data-sniffer probes fused with the worlds' most used sites.

As an addon architect soon, my projected expectations for a good platform, APIs, and Addons are;
-FF as rock-solid platform
-APIs as an interface that prevents conflicts FF-addons-OS as much as possible
-Addons that followthe guidelines, but with a rock-solid API can be written by a rational people and not to go OCD if accidental omission will stall or crash the browser.

I can foresee some general solutions to addons incompatibility problems, alas too early to write before looking under the hood phase.

Will report more, hopefully sort out what causes messages not passing reliably from KeepassXC out to the native app and vice-versa, as I dig deeper into the addons debugging and architecture.

Best,
